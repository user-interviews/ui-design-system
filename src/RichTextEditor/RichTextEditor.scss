@import '../../scss/theme.scss';

.RichTextEditor {
  width: 100%;

  .RichTextEditor__character-count {
    @include synth-font-type-20;
    margin-top: 0.5rem;
  }
}


.RichTextEditor__field--error {
  .ProseMirror {
    border-color: $ux-red-600;
  }
}

.ProseMirror {
  min-height: 5rem;
  background: $ux-neutral-100;
  border: 1px solid $ux-gray-400;
  border-radius: 0 0 $ux-border-radius $ux-border-radius;
  padding: 1rem;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;

  &:focus {
    border: 0.06rem solid $ux-blue-500;
    box-shadow: $input-focus-box-shadow;
    color: $input-focus-color;
    filter: drop-shadow(0px 0px 3px rgba(102, 175, 233, 0.7));
    outline: none;
  }

  > * + * {
    margin-top: 0.75em;
  }

  ul,
  ol {
    padding: 0 1rem;
  }

  li > p {
    margin: 0;
  }

  p.is-editor-empty:first-child::before {
    color: $ux-gray-700;
    content: attr(data-placeholder);
    float: left;
    height: 0;
    pointer-events: none;
  }
}
