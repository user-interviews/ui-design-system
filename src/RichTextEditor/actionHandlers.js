export const createActionHandlers = (editor) => ({
  link: () => {
    // TODO: use DS components for link prompt
    const previousUrl = editor.getAttributes('link').href;
    const url = window.prompt('URL', previousUrl);

    // cancelled
    if (url === null) {
      return;
    }

    // empty
    if (url === '') {
      editor.chain().focus().extendMarkRange('link').unsetLink()
        .run();

      return;
    }

    const protocolRegex = /^http(s?):\/\//;
    const absoluteUrl = url.match(protocolRegex) ? url : `//${url}`;

    // update link
    editor.chain().focus().extendMarkRange('link').setLink({ href: absoluteUrl })
      .run();
  },
  unlink: () => editor.chain().focus().unsetLink().run(),
  bold: () => editor.chain().focus().toggleBold().run(),
  italic: () => editor.chain().focus().toggleItalic().run(),
  unorderedList: () => editor.chain().focus().toggleBulletList().run(),
  orderedList: () => editor.chain().focus().toggleOrderedList().run(),
});
