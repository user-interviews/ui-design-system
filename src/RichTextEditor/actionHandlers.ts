import type { Editor } from '@tiptap/core';

export const createActionHandlers = (editor: Editor) => ({
  link: () => {
    // TODO: use DS components for link prompt
    const previousUrl = editor.getAttributes('link').href;
    const url = window.prompt('URL', previousUrl);

    // cancelled
    if (url === null) {
      return false;
    }

    // empty
    if (url === '') {
      editor
        .chain()
        .focus()
        .extendMarkRange('link')
        .unsetLink()
        .run();

      return false;
    }

    const protocolRegex = /^http(s?):\/\//;
    const absoluteUrl = url.match(protocolRegex) ? url : `//${url}`;

    // update link
    editor
      .chain()
      .focus()
      .extendMarkRange('link')
      .setLink({ href: absoluteUrl })
      .run();

    return false;
  },
  unlink: () => editor.chain().focus().unsetLink().run(),
  bold: () => editor.chain().focus().toggleBold().run(),
  italic: () => editor.chain().focus().toggleItalic().run(),
  unorderedList: () => editor.chain().focus().toggleBulletList().run(),
  orderedList: () => editor.chain().focus().toggleOrderedList().run(),
});
