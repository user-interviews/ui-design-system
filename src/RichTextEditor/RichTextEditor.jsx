import './RichTextEditor.scss';

import React from 'react';
import * as propTypes from 'prop-types';

import classNames from 'classnames';

import { EditorContent, useEditor } from '@tiptap/react';

import Bold from '@tiptap/extension-bold';
import BulletList from '@tiptap/extension-bullet-list';
import CharacterCount from '@tiptap/extension-character-count';
import Document from '@tiptap/extension-document';
import HardBreak from '@tiptap/extension-hard-break';
import History from '@tiptap/extension-history';
import Italic from '@tiptap/extension-italic';
import Link from '@tiptap/extension-link';
import ListItem from '@tiptap/extension-list-item';
import OrderedList from '@tiptap/extension-ordered-list';
import Paragraph from '@tiptap/extension-paragraph';
import Placeholder from '@tiptap/extension-placeholder';
import Text from '@tiptap/extension-text';

import sanitizeHtml from 'sanitize-html';

import { LoadingSkeleton } from 'src/LoadingSkeleton';
import { TemplateVariable, buildSuggestions } from './TemplateVariable';

import RichTextEditorMenuBar from './RichTextEditorMenuBar';

import { OneLineLimit } from './oneLineLimit';

import { RichTextEditorActions, RichTextEditorAllActionsArray, RichTextEditorDefaultActionsArray } from './richTextEditorActions';
import { createActionHandlers } from './actionHandlers';

const ExtendedLink = Link.extend({
  addKeyboardShortcuts() {
    const actionHandlers = createActionHandlers(this.editor);

    return {
      'Mod-k': actionHandlers.link,
    };
  },
});

const RichTextEditor = ({
  allowedAttributes,
  allowedTags,
  ariaAttributes,
  availableActions,
  characterLimit,
  className,
  hasErrors,
  id,
  initialValue,
  isOneLine,
  onChange,
  placeholder,
  templateVariables,
}) => {
  const oneLineExtension = isOneLine ? [OneLineLimit] : [];
  const hasTemplateVariables = templateVariables.length > 0;

  const requiredExtensions = [
    Document,
    Text,
    History,
    HardBreak,
    Paragraph,
    ListItem,
    Placeholder.configure({
      placeholder,
    }),
    CharacterCount.configure({
      limit: characterLimit,
    }),
  ];

  const templateVariablesExtension = hasTemplateVariables ?
    [TemplateVariable.configure({
      HTMLAttributes: {
        class: 'RichTextEditor__TemplateVariable',
      },
      suggestion: buildSuggestions(templateVariables),
    })] : [];

  const optionalExtensions = [
    {
      name: RichTextEditorActions.BOLD,
      config: Bold,
    },
    {
      name: RichTextEditorActions.ITALIC,
      config: Italic,
    },
    {
      name: RichTextEditorActions.LINK,
      config: ExtendedLink,
    },
    {
      name: RichTextEditorActions.UNORDERED_LIST,
      config: BulletList,
    },
    {
      name: RichTextEditorActions.ORDERED_LIST,
      config: OrderedList,
    },
  ].filter((extension) => availableActions.includes(extension.name))
    .map((extension) => extension.config);

  const extensions = [
    ...oneLineExtension,
    ...requiredExtensions,
    ...templateVariablesExtension,
    ...optionalExtensions,
  ];

  const editor = useEditor({
    extensions,
    content: initialValue,
    onUpdate: ({ editor: ttEditor }) => {
      const html = ttEditor.getHTML();

      // if allowAttributes or allowedTags aren't passed
      // then use defaults from sanitize-html by not passing that key in the options
      // https://github.com/apostrophecms/sanitize-html

      const options = {};

      if (allowedAttributes) {
        options.allowedAttributes = allowedAttributes;
      }

      if (allowedTags) {
        options.allowedTags = allowedTags;
      }

      // When using template variables, we need to whitelist some specific attributes so that
      // the editor can re-initialize correctly from db state
      if (hasTemplateVariables) {
        options.allowedAttributes = {
          ...(options.allowedAttributes || {}),
          span: ['class', 'data-id', 'data-type'],
        };

        if (options.allowedTags && !options.allowedTags.includes('span')) {
          options.allowedTags.push(['span']);
        }
      }

      const sanitizedHtml = sanitizeHtml(html, options);

      onChange(sanitizedHtml);
    },
  });

  return (
    editor ? (
      <div
        className="RichTextEditor"
        id={id}
      >
        {availableActions.length > 0 && (
          <RichTextEditorMenuBar
            availableActions={availableActions}
            editor={editor}
          />
        )}
        <EditorContent
          className={classNames(
            className,
            'RichTextEditor__field',
            { 'RichTextEditor__field--error': hasErrors },
            { 'RichTextEditor__field--without-menu-bar': availableActions.length === 0 },
          )}
          editor={editor}
          role="textbox"
          {...ariaAttributes}
        />
        {
          !!characterLimit && (
            <p className="RichTextEditor__character-count">
              {editor.storage.characterCount.characters()}/{characterLimit}
            </p>
          )
        }
      </div>
    ) : (
      <>
        <LoadingSkeleton height={40} />
        <LoadingSkeleton height={70} />
      </>
    )
  );
};

RichTextEditor.propTypes = {
  /**
   HTML attributes to allow while sanitizing the editor's content.
   If none provided then it uses sanitize-html's defaults
   https://github.com/apostrophecms/sanitize-html#default-options
  */
  allowedAttributes: propTypes.array,
  /**
   HTML tags to allow while sanitizing the editor's content.
   If none provided then it uses sanitize-html's defaults
   https://github.com/apostrophecms/sanitize-html#default-options
  */
  allowedTags: propTypes.array,
  /**
    ARIA attributes to pass along to the text editor field
    for accessibility purposes. The keys of the object should be valid
    ARIA attributes.
  */
  ariaAttributes: propTypes.object,
  /**
   Which actions to include in the taskbar. Available as constants.
   Current options are BOLD, ITALIC, LINK, UNLINK, UNORDERED_LIST and ORDERED_LIST
  */
  availableActions: propTypes.arrayOf(propTypes.oneOf(RichTextEditorAllActionsArray)),
  characterLimit: propTypes.number,
  className: propTypes.string,
  hasErrors: propTypes.bool,
  id: propTypes.string.isRequired,
  /**
   HTML provided to initialize the editor's content
  */
  initialValue: propTypes.string,
  isOneLine: propTypes.bool,
  placeholder: propTypes.string,
  templateVariables: propTypes.arrayOf(propTypes.string),
  /**
    Callback function to call with sanitized HTML when editor state changes
  */
  onChange: propTypes.func.isRequired,
};

RichTextEditor.defaultProps = {
  allowedAttributes: undefined,
  allowedTags: undefined,
  ariaAttributes: undefined,
  availableActions: RichTextEditorDefaultActionsArray,
  characterLimit: undefined,
  className: undefined,
  hasErrors: false,
  initialValue: undefined,
  isOneLine: false,
  placeholder: undefined,
  templateVariables: [],
};

export default RichTextEditor;
