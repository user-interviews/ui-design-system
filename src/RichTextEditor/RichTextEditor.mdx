import { ArgTypes,  Canvas } from '@storybook/blocks';
import RichTextEditor from './RichTextEditor';
import * as ComponentStories from './RichTextEditor.stories';

# RichTextEditor

<h4>
  RichTextEditor is a field element which provides a "what you see is what you get" (WYSIWYG) editor for HTML content.
</h4>

<Canvas of={ComponentStories.Default} />

### When to use
To allow users to edit text where formatting is important, such as messages to be sent via email.

### When to not use
When formatting is not important. In those cases regular form inputs are a better fit.

## Props

<ArgTypes of={RichTextEditor} />

## Stories

By default RichTextEditor includes all available formatting actions.
If you'd like to limit which options are available you can do so by passing the actions you'd like to the `availableActions` prop in an array.

<Canvas of={ComponentStories.AvailableActions} />

Provide an object where the keys are ARIA attributes to `ariaAttributes` to pass them along to the text input field for accessibility purposes.

<Canvas of={ComponentStories.ARIAAttributes} />

If you set a character limit then the length of the inputted text will be restricted accordingly, including if longer text is pasted.
A counter will also be shown below the input so you know how close you are to hitting the limit.

<Canvas of={ComponentStories.CharacterLimit} />

You are also able to limit input to one line. This can be useful in situations where a limited amount of text is idea, but formatting is important.
The `isOneLine` and `characterLimit` props can be used together in cases where you really need to restrict text length.

<Canvas of={ComponentStories.OneLine} />

RichTextEditor also has an error state to display that the input is invalid. The most common use case for this is when it is a required field and the input is empty,
but by setting `hasErrors` via your provided `onChange` callback you can check the HTML produced by the RichTextEditor for any requirements that you have.

<Canvas of={ComponentStories.Error} />

RichTextEditor can optionally take in a forwardedRef to allow content to be programmatically manipulated.

<Canvas of={ComponentStories.SetContent} />

RichTextEditor can be set to read-only mode by setting the `editable` prop to `false` (e.g., when user permissions don't allow them to edit a certain field).

<Canvas of={ComponentStories.NotEditable} />

RichTextEditor can be configured to be used in display mode only which is useful when displaying rich text when you have to do something extra with it (e.g character truncation with summary display).
Modify `displayMode` to `true`. This will automatically put the editor in not editable mode as well. You can also provide optional `onTruncate` callback that will be called if RichTextEditor truncates text.

<Canvas of={ComponentStories.DisplayMode} />

If you're using RichTextEditor with `displayMode` and `characterLimit` then your text will be truncated to plaintext if it's above the char limit.
But if your text is below the limit, then it'll still be displayed as rich text.

E.g with text above the `characterLimit`

<Canvas of={ComponentStories.DisplayModeWithCharacterLimitAbove} />

E.g with text below the `characterLimit`

<Canvas of={ComponentStories.DisplayModeWithCharacterLimitBelow} />

