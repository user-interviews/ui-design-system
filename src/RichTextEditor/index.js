import RichTextEditor from './RichTextEditor';
import { AllRteActions, DefaultRteActions, RteActions } from './richTextEditorActions';

export {
  AllRteActions,
  DefaultRteActions,
  RichTextEditor,
  RteActions,
};
