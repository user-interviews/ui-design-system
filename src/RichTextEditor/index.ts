import RichTextEditor from './RichTextEditor';
import { RichTextEditorAllActionsArray, RichTextEditorDefaultActionsArray, RichTextEditorActions } from './richTextEditorActions';

export {
  RichTextEditorAllActionsArray,
  RichTextEditorDefaultActionsArray,
  RichTextEditor,
  RichTextEditorActions,
};
