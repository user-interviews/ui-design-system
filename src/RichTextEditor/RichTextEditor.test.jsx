import React from 'react';
import { render, screen } from '@testing-library/react';

import RichTextEditor from './RichTextEditor';

describe('<RichTextEditor />', () => {
  const Setup = (overrides = {}) => (
    <RichTextEditor
      id="some-id"
      onChange={jest.fn()}
      {...overrides}
    />
  );

  it('renders snapshot', () => {
    const { asFragment } = render(<Setup />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('given an initial value', () => {
    it('deserializes value correctly', () => {
      render(<Setup initialValue="<p>hello world</p>" />);

      expect(screen.getByText('hello world')).toBeInTheDocument();
    });
  });
});
