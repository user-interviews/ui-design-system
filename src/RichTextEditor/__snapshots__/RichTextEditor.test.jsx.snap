// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RichTextEditor /> renders snapshot 1`] = `
<DocumentFragment>
  <div
    class="RichTextEditor"
    id="some-id"
  >
    <div
      class="RichTextEditorMenuBar"
    >
      <button
        aria-label="Bold"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-bold fa-fw"
          data-icon="bold"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
<<<<<<< release/2.1.1
            d="M0 56C0 42.7 10.7 32 24 32H72 88 212c68.5 0 124 55.5 124 124c0 34.7-14.3 66.2-37.3 88.7C339.7 264.9 368 307.1 368 356c0 68.5-55.5 124-124 124H88 72 24c-13.3 0-24-10.7-24-24s10.7-24 24-24H48V256 80H24C10.7 80 0 69.3 0 56zM212 232c42 0 76-34 76-76s-34-76-76-76H96V232H212zM96 280V432H244c42 0 76-34 76-76s-34-76-76-76H212 96z"
=======
            d="M314.52 238.78A119.76 119.76 0 0 0 232 32H48a16 16 0 0 0-16 16v16a16 16 0 0 0 16 16h16v352H48a16 16 0 0 0-16 16v16a16 16 0 0 0 16 16h208a128 128 0 0 0 128-128c0-49.49-28.38-91.92-69.48-113.22zM128 80h88a72 72 0 0 1 0 144h-88zm112 352H128V272h112a80 80 0 0 1 0 160z"
>>>>>>> main
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Italic"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-italic fa-fw"
          data-icon="italic"
          data-prefix="far"
          focusable="false"
          role="img"
<<<<<<< release/2.1.1
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M128 56c0-13.3 10.7-24 24-24H360c13.3 0 24 10.7 24 24s-10.7 24-24 24H291.3L144.7 432H232c13.3 0 24 10.7 24 24s-10.7 24-24 24H24c-13.3 0-24-10.7-24-24s10.7-24 24-24H92.7L239.3 80H152c-13.3 0-24-10.7-24-24z"
=======
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 48v16a16 16 0 0 1-16 16h-67l-88 352h59a16 16 0 0 1 16 16v16a16 16 0 0 1-16 16H16a16 16 0 0 1-16-16v-16a16 16 0 0 1 16-16h67l88-352h-59a16 16 0 0 1-16-16V48a16 16 0 0 1 16-16h192a16 16 0 0 1 16 16z"
>>>>>>> main
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Link"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-link fa-fw"
          data-icon="link"
          data-prefix="far"
          focusable="false"
          role="img"
<<<<<<< release/2.1.1
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M580.3 267.2c56.2-56.2 56.2-147.3 0-203.5C526.8 10.2 440.9 7.3 383.9 57.2l-6.1 5.4c-10 8.7-11 23.9-2.3 33.9s23.9 11 33.9 2.3l6.1-5.4c38-33.2 95.2-31.3 130.9 4.4c37.4 37.4 37.4 98.1 0 135.6L433.1 346.6c-37.4 37.4-98.2 37.4-135.6 0c-35.7-35.7-37.6-92.9-4.4-130.9l4.7-5.4c8.7-10 7.7-25.1-2.3-33.9s-25.1-7.7-33.9 2.3l-4.7 5.4c-49.8 57-46.9 142.9 6.6 196.4c56.2 56.2 147.3 56.2 203.5 0L580.3 267.2zM59.7 244.8C3.5 301 3.5 392.1 59.7 448.2c53.6 53.6 139.5 56.4 196.5 6.5l6.1-5.4c10-8.7 11-23.9 2.3-33.9s-23.9-11-33.9-2.3l-6.1 5.4c-38 33.2-95.2 31.3-130.9-4.4c-37.4-37.4-37.4-98.1 0-135.6L207 165.4c37.4-37.4 98.1-37.4 135.6 0c35.7 35.7 37.6 92.9 4.4 130.9l-5.4 6.1c-8.7 10-7.7 25.1 2.3 33.9s25.1 7.7 33.9-2.3l5.4-6.1c49.9-57 47-142.9-6.5-196.5c-56.2-56.2-147.3-56.2-203.5 0L59.7 244.8z"
=======
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M314.222 197.78c51.091 51.091 54.377 132.287 9.75 187.16-6.242 7.73-2.784 3.865-84.94 86.02-54.696 54.696-143.266 54.745-197.99 0-54.711-54.69-54.734-143.255 0-197.99 32.773-32.773 51.835-51.899 63.409-63.457 7.463-7.452 20.331-2.354 20.486 8.192a173.31 173.31 0 0 0 4.746 37.828c.966 4.029-.272 8.269-3.202 11.198L80.632 312.57c-32.755 32.775-32.887 85.892 0 118.8 32.775 32.755 85.892 32.887 118.8 0l75.19-75.2c32.718-32.725 32.777-86.013 0-118.79a83.722 83.722 0 0 0-22.814-16.229c-4.623-2.233-7.182-7.25-6.561-12.346 1.356-11.122 6.296-21.885 14.815-30.405l4.375-4.375c3.625-3.626 9.177-4.594 13.76-2.294 12.999 6.524 25.187 15.211 36.025 26.049zM470.958 41.04c-54.724-54.745-143.294-54.696-197.99 0-82.156 82.156-78.698 78.29-84.94 86.02-44.627 54.873-41.341 136.069 9.75 187.16 10.838 10.838 23.026 19.525 36.025 26.049 4.582 2.3 10.134 1.331 13.76-2.294l4.375-4.375c8.52-8.519 13.459-19.283 14.815-30.405.621-5.096-1.938-10.113-6.561-12.346a83.706 83.706 0 0 1-22.814-16.229c-32.777-32.777-32.718-86.065 0-118.79l75.19-75.2c32.908-32.887 86.025-32.755 118.8 0 32.887 32.908 32.755 86.025 0 118.8l-45.848 45.84c-2.93 2.929-4.168 7.169-3.202 11.198a173.31 173.31 0 0 1 4.746 37.828c.155 10.546 13.023 15.644 20.486 8.192 11.574-11.558 30.636-30.684 63.409-63.457 54.733-54.735 54.71-143.3-.001-197.991z"
>>>>>>> main
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-disabled="true"
        aria-label="Unlink"
        class="Button IconButton btn btn-transparent btn-sm"
        disabled=""
        type="button"
      >
        <svg
          aria-hidden="true"
<<<<<<< release/2.1.1
          class="svg-inline--fa fa-link-slash fa-fw"
          data-icon="link-slash"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7L489.3 358.2l91-91c56.2-56.2 56.2-147.3 0-203.5C526.8 10.2 440.9 7.3 383.9 57.2l-6.1 5.4c-10 8.7-11 23.9-2.3 33.9s23.9 11 33.9 2.3l6.1-5.4c38-33.2 95.2-31.3 130.9 4.4c37.4 37.4 37.4 98.1 0 135.6l-95.1 95.1-45.5-35.7c24.2-53.3 14.1-117.9-29.3-161.3c-52.1-52.1-134.4-55.9-190.8-11.2L38.8 5.1zm186.3 146c36.9-22.3 85.6-17.6 117.4 14.3c26 26 34 63.3 23.7 96.4L225.1 151.1zM406.9 416.6l-54.2-42.7c-20.2-2.7-39.7-11.7-55.2-27.3c-9.8-9.8-17-21.2-21.7-33.3l-54.2-42.7c-2.2 39.6 11.9 79.9 41.9 109.9c38.8 38.8 94.2 50.8 143.4 36zM116.6 187.9L59.7 244.8C3.5 301 3.5 392.1 59.7 448.2c53.6 53.6 139.5 56.4 196.5 6.5l6.1-5.4c10-8.7 11-23.9 2.3-33.9s-23.9-11-33.9-2.3l-6.1 5.4c-38 33.2-95.2 31.3-130.9-4.4c-37.4-37.4-37.4-98.1 0-135.6l60.9-60.9-38-29.9z"
=======
          class="svg-inline--fa fa-unlink fa-fw"
          data-icon="unlink"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M304.083 388.936c4.686 4.686 4.686 12.284 0 16.971l-65.057 65.056c-54.709 54.711-143.27 54.721-197.989 0-54.713-54.713-54.719-143.27 0-197.989l65.056-65.057c4.686-4.686 12.284-4.686 16.971 0l22.627 22.627c4.686 4.686 4.686 12.284 0 16.971L81.386 311.82c-34.341 34.341-33.451 88.269.597 120.866 32.577 31.187 84.788 31.337 117.445-1.32l65.057-65.056c4.686-4.686 12.284-4.686 16.971 0l22.627 22.626zm-56.568-243.245l64.304-64.304c34.346-34.346 88.286-33.453 120.882.612 31.18 32.586 31.309 84.785-1.335 117.43l-65.056 65.057c-4.686 4.686-4.686 12.284 0 16.971l22.627 22.627c4.686 4.686 12.284 4.686 16.971 0l65.056-65.057c54.711-54.709 54.721-143.271 0-197.99-54.71-54.711-143.27-54.72-197.989 0l-65.057 65.057c-4.686 4.686-4.686 12.284 0 16.971l22.627 22.627c4.685 4.685 12.283 4.685 16.97-.001zm238.343 362.794l22.627-22.627c4.686-4.686 4.686-12.284 0-16.971L43.112 3.515c-4.686-4.686-12.284-4.686-16.971 0L3.515 26.142c-4.686 4.686-4.686 12.284 0 16.971l465.373 465.373c4.686 4.686 12.284 4.686 16.97-.001z"
>>>>>>> main
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Unordered List"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-list-ul fa-fw"
          data-icon="list-ul"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
<<<<<<< release/2.1.1
            d="M64 64a32 32 0 1 0 0 64 32 32 0 1 0 0-64zm120 8c-13.3 0-24 10.7-24 24s10.7 24 24 24H488c13.3 0 24-10.7 24-24s-10.7-24-24-24H184zm0 160c-13.3 0-24 10.7-24 24s10.7 24 24 24H488c13.3 0 24-10.7 24-24s-10.7-24-24-24H184zm0 160c-13.3 0-24 10.7-24 24s10.7 24 24 24H488c13.3 0 24-10.7 24-24s-10.7-24-24-24H184zM96 256a32 32 0 1 0 -64 0 32 32 0 1 0 64 0zM64 384a32 32 0 1 0 0 64 32 32 0 1 0 0-64z"
=======
            d="M48 368a48 48 0 1 0 48 48 48 48 0 0 0-48-48zm0-160a48 48 0 1 0 48 48 48 48 0 0 0-48-48zm0-160a48 48 0 1 0 48 48 48 48 0 0 0-48-48zm448 24H176a16 16 0 0 0-16 16v16a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16V88a16 16 0 0 0-16-16zm0 160H176a16 16 0 0 0-16 16v16a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-16a16 16 0 0 0-16-16zm0 160H176a16 16 0 0 0-16 16v16a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-16a16 16 0 0 0-16-16z"
>>>>>>> main
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Ordered List"
        class="Button IconButton btn btn-transparent btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-list-ol fa-fw"
          data-icon="list-ol"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
<<<<<<< release/2.1.1
            d="M24 56c0-13.3 10.7-24 24-24H80c13.3 0 24 10.7 24 24V176h16c13.3 0 24 10.7 24 24s-10.7 24-24 24H40c-13.3 0-24-10.7-24-24s10.7-24 24-24H56V80H48C34.7 80 24 69.3 24 56zM86.7 341.2c-6.5-7.4-18.3-6.9-24 1.2L51.5 357.9c-7.7 10.8-22.7 13.3-33.5 5.6s-13.3-22.7-5.6-33.5l11.1-15.6c23.7-33.2 72.3-35.6 99.2-4.9c21.3 24.4 20.8 60.9-1.1 84.7L86.8 432H120c13.3 0 24 10.7 24 24s-10.7 24-24 24H32c-9.5 0-18.2-5.6-22-14.4s-2.1-18.9 4.3-25.9l72-78c5.3-5.8 5.4-14.6 .3-20.5zM216 72H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm0 160H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm0 160H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
=======
            d="M61.77 401l17.5-20.15a19.92 19.92 0 0 0 5.07-14.19v-3.31C84.34 356 80.5 352 73 352H16a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h22.84a154.82 154.82 0 0 0-11 12.31l-5.61 7c-4 5.07-5.25 10.13-2.8 14.88l1.05 1.93c3 5.76 6.3 7.88 12.25 7.88h4.73c10.33 0 15.94 2.44 15.94 9.09 0 4.72-4.2 8.22-14.36 8.22a41.54 41.54 0 0 1-15.47-3.12c-6.49-3.88-11.74-3.5-15.6 3.12l-5.59 9.31c-3.73 6.13-3.2 11.72 2.62 15.94 7.71 4.69 20.39 9.44 37 9.44 34.16 0 48.5-22.75 48.5-44.12-.03-14.38-9.12-29.76-28.73-34.88zM496 392H176a16 16 0 0 0-16 16v16a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-16a16 16 0 0 0-16-16zm0-320H176a16 16 0 0 0-16 16v16a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16V88a16 16 0 0 0-16-16zm0 160H176a16 16 0 0 0-16 16v16a16 16 0 0 0 16 16h320a16 16 0 0 0 16-16v-16a16 16 0 0 0-16-16zM16 160h64a8 8 0 0 0 8-8v-16a8 8 0 0 0-8-8H64V40a8 8 0 0 0-8-8H32a8 8 0 0 0-7.14 4.42l-8 16A8 8 0 0 0 24 64h8v64H16a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8zm-3.9 160H80a8 8 0 0 0 8-8v-16a8 8 0 0 0-8-8H41.33c3.28-10.29 48.33-18.68 48.33-56.44 0-29.06-25-39.56-44.47-39.56-21.36 0-33.8 10-40.45 18.75-4.38 5.59-3 10.84 2.79 15.37l8.58 6.88c5.61 4.56 11 2.47 16.13-2.44a13.4 13.4 0 0 1 9.45-3.84c3.33 0 9.28 1.56 9.28 8.75C51 248.19 0 257.31 0 304.59v4C0 316 5.08 320 12.1 320z"
>>>>>>> main
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="RichTextEditor__field"
      role="textbox"
    >
      <div
        class="ProseMirror"
        contenteditable="true"
        tabindex="0"
        translate="no"
      >
        <p
          class="is-empty is-editor-empty"
        >
          <br
            class="ProseMirror-trailingBreak"
          />
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
